### DISCUSSION TITLE
Fetching Dynamic Data for a client component through a server component.

### DISCUSSION THREAD

**Discussion by Nicklas-17** [vercel/next.js#64550] [App Router]
Hi Guys, hope you’re well.  I've been struggling with the following on Nextjs and would really appreciate your input : I’ve setup a dynamic client component that fetches data from an External API works well on my local host.

But when I deploy it on GitHub pages I get CORS error.

Research says I got to setup a route handler - make the fetch pass through a server component like a proxy.

Struggling to setup the route handler. It’s not working. Below is my client component code - would appreciate it if I could get the actual code answer. Still pretty new to nextjs :  here is my code (ps im putting my code in the code brackets but doesn't seem to mark as code..) : 

`

use client;

import React, { useState, useEffect } from "react";
import styles from "./ProductTray.module.css";

export default function ProductTray() {
  const [products, setProducts] = useState([]);
  const [currentProductIndex, setCurrentProductIndex] = useState(0);
  const [currentPage, setCurrentPage] = useState(1);

  const handlePageChange = (page) => {
    setCurrentPage(page);
    setCurrentProductIndex(0);
  };

 
  useEffect(function fetchData() {
    fetch(
      "externalapiurl'',
      {
        headers: {
          Authorization:
            " Bearer .....
          Accept: ".....",
          "content-Type": "....",
        },
      }
    )
      .then((response) => response.json())

      .then((data) => setProducts(data.data))
      .catch((error) => console.error("Error fetching data", error));
  }, []);

  const handleKeyDown = (event) => {
    if (event.key === "ArrowLeft") {
      if (currentProductIndex > 0) {
        setCurrentProductIndex(currentProductIndex - 1);
      } else if (currentPage > 1) {
        handlePageChange(currentPage - 1);
        setCurrentProductIndex(products.length - 1);
      }
    } else if (event.key === "ArrowRight") {
      if (currentProductIndex < products.length - 1) {
        setCurrentProductIndex(currentProductIndex + 1);
      } else {
        handlePageChange(currentPage + 1);
        setCurrentProductIndex(0);
      }
    }
  };

  useEffect(() => {
    document.addEventListener("keydown", handleKeyDown);

    return () => {
      document.removeEventListener("keydown", handleKeyDown);
    };
  }, [currentProductIndex, currentPage]);

  const currentProduct = products[currentProductIndex];

  return (
    <>
      {currentProduct && ( // Conditionally render only if product exists
        <div key={currentProduct.id}>
          {" "}
          <div className={styles.productImages}>
            {" "}
            {currentProduct.photos?.slice(0, 4).map(
              (
                image,
                index // Limit to 4 images
              ) => (
                <img
                  key={index} // Key for each image
                  src={image.photoUrl}
                  alt={`Product Image ${index + 1}`}
                  className={styles.productImage}
                />
              )
            )}
          </div>
          {/* Other product details */}
        </div>
      )}
      <div className={styles.pagination}>
        {/* ... your pagination controls here ... */}
      </div>
    </>
  );
}
`

**Comment by Icegreeen** [comment]
Instead of making the request directly on the client side, you will make the request through your Next.js server using a proxy.

You can do this by creating a file in the pages/api folder with code that makes a request to an external API and forwards a response back to the client. Next, you need to update the fetch request URL in your client component to point to this Next.js server endpoint. See an example of how you can do this.

```
// pages/api/externalapi.js

export default async function handler(req, res) {
  try {
    const response = await fetch('externalapiurl', {
      method: req.method,
      headers: {
        ...req.headers,
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      throw new Error('Failed to fetch data from external API');
    }

    const data = await response.json();
    res.status(200).json(data);
  } catch (error) {
    console.error('Error fetching data from external API:', error);
    res.status(500).json({ error: 'Failed to fetch data from external API' });
  }
}

```

Modify the request code in your client application:

```
useEffect(function fetchData() {
  fetch('/api/externalapi', { method: 'GET' })
    .then((response) => response.json())
    .then((data) => setProducts(data.data))
    .catch((error) => console.error('Error fetching data', error));
}, []);

```

This example will allow your client side requests to pass through the Next.js server, solving your CORS problem.

@Nicklas-17

**Reply by Nicklas-17** [reply]
Thank you for taking the time to answer, I have indeed tried this and I am getting : [Error] Failed to load resource: the server responded with a status of 405 (Method Not Allowed)  any ideas on why that might be happening ?

**Reply by Icegreeen** [reply]
According to the error you must be using the incorrect method, which method are you using on the client side? get, put, delete or patch ?

and which method does the route accept?

**Reply by Nicklas-17** [reply]
GET on both if I am not mistaken : here is the code for the client : ` useEffect(() => {
    async function fetchData() {
      const res = await fetch("/api/products", { method: "GET" });
      const data = await res.json();
      setProducts(data.data);
    }
    fetchData();
  }, []);`   and here Is the code for the route : `import { NextResponse } from "next/server";

export default async function handler(req, res) {
  try {
    const response = await fetch(
      "external URL",
      {
        headers: {
          Authorization:
            "Bearer  .....",
          Accept: "application/json",
          "Content-Type": "application/json",
        },
      }
    );
    const data = await response.json();
    res.status(200).json(data);
  } catch (error) {
    console.error("Error fetching data from external API:", error);
    res.status(500).json({ error: "Failed to fetch data from external API" });
  }
}

export { handler };
`

**Comment by icyJoseph** [comment]
Hi,

The problem here is two fold:

- GitHub Pages only support static sites, that is, it is more or less a CDN with extra features, that serves your HTML, JS and CSS (among other static assets), which means there's no server that could handle, requests from your client to proxy further up to the external API
  - Funnily enough, since some Route Handlers might be statically optimized, they'll be available in a static site deploy. How? Well the static optimization creates a JSON file with the data produced by the Route Handler at build time, and that's a static asset that can be served 
- Your external API needs to allow cross origin requests (from your GitHub Pages domain), and for that you have to make changes to the external API ~ This is not necessarily something you can control from the browser, it is there as a protection mechanism, enforced by the browser. You can lower the bar, but then the data won't be accessible. Do you have control over this external API?

**Reply by Nicklas-17** [reply]
Hi Joseph, thanks for your answer. I don't have access to the external API. How can I optimise the route handler in your first solution proposal ?

**Reply by icyJoseph** [reply]
I think you might have the wrong design then. Who owns this external API? Can you ask them to allow your GitHub pages domain?

IF this data is pretty much static, then during build time, given some conditions, Next.js statically optimizes the request and outputs a JSON file instead, which your client could try to fetch. Again, it'll be static!

Read here to get those "conditions" in place: https://nextjs.org/docs/app/building-your-application/routing/route-handlers#caching

I notice that you posted an API Route in your other comments, those won't be optimized. It has to be Route Handlers, in App Router.

**Reply by Nicklas-17** [reply]
My client component is displaying/fetching based on user interaction ( so not static data ) -  hence if I understand correctly what you're saying : it is not possible to have this working on GitHub pages unless the external API allows  cross origin requests (from my GitHub Pages domain) ?

**Comment by obaidshah345** [comment]
Hey, I know it's pretty old but I've found out that this is still a known issue for folks on Nextjs 13.x.x and beyond using API route handlers.

So, If you are using distDir: "build" in your Nextjs config because of a 500 Error on API requests when using the output: "export" on localhost then you need to make one change in the Vercel project settings before you deploy to Vercel.

Open your "Project Settings" and go to "Build & Development Settings" on Vercel, now check the Output Directory field and see if it contains a ".next" in there, if yes, then you need to override it and change it to "build" as this is what you are currently building your static assets to on localhost. Save your settings and deploy.

Let me know how it goes. 

Ciao,
Sye
